src/abstractdataframe/selection.jl

```diff
@@ -0,0 +1,424 @@
+    else
+        newcol = Symbol(join(_names(idx)[c], "_"), "_", funname(fun))
+    end
+    return c => fun => newcol
+end
+
+function select_transform!(nc::Union{Pair{Int, Pair{ColRename, Symbol}},
+                                     Pair{<:Union{Int, AbstractVector{Int}},
+                                          <:Pair{<:Base.Callable, Symbol}}},
+                           df::DataFrame, newdf::DataFrame,
+                           transformed_cols::Dict{Symbol, Any}, copycols::Bool)
+    col_idx = first(nc)
+    transform_spec = last(nc)
+    newname = last(transform_spec)
+    if !isnothing(transformed_cols[newname])
+        @assert !hasproperty(newdf, newname)
+    end
+    if nc isa Pair{Int, Pair{ColRename, Symbol}}
+        newdf[!, newname] = copycols ? df[:, col_idx] : df[!, col_idx]
+    elseif nc isa Pair{Int, <:Pair{<:Base.Callable, Symbol}}
+        newdf[!, newname] = first(transform_spec).(df[!, col_idx])
+    elseif nc isa Pair{<:AbstractVector{Int}, <:Pair{<:Base.Callable, Symbol}}
+        if length(col_idx) == 0
+            newdf[!, newname] = [first(transform_spec)() for _ in axes(df, 1)]
+        else
+            rowiterator = Tables.rows(Tables.columntable(df[!, col_idx]))
+            newdf[!, newname] = map(first(transform_spec),
+                                    Tables.namedtupleiterator(eltype(rowiterator), rowiterator))
+        end
+    else
+        throw(ErrorException("code should never reach this branch"))
+    end
+    transformed_cols[newname] = nothing
+end
+
+"""
+    select!(df::DataFrame, inds...)
+
+Mutate `df` in place to retain only columns specified by `inds...` and return it.
+
+Arguments passed as `inds...` can be any index that is allowed for column indexing
+provided that the columns requested in each of them are unique and present in `df`.
+In particular, regular expressions, `All`, `Between`, and `Not` selectors are supported.
+
+Column renaming and transformations are supported.
+The syntax for column renaming is `old_column=>new_column_name`.
+The syntax for column transformations is `old_column=>fun=>new_column_name`.
+`new_column_name` must be a `Symbol`.
+If `old_column` is a `Symbol` or an integer then and `fun` must be callable
+that is applied to each element (row) of `old_column`.
+Otherwise `old_column` can be any column indexing syntax, but in this case `fun`
+will be passed a `NamedTuple` representing each row, holding only
+the columns specified by `old_column`.
+
+Column transfomation syntax also supports a short `old_column=>fun` form, in which
+case `new_column_name` is automatically generated by joining `old_column` with `fun` name
+with `_`.
+
+If a collection of column names is passed to `select` then requesting duplicate column names
+in target data frame are accepted (e.g. `select(df, [:a], :, r"a")` is allowed)
+and they get deduplicated.
+The exception is for single column selection, column renaming and column transformation
```

Comment:
This isn't an exception AFAICT. I suggest this:
```
In particular a syntax to move column `:col` to the first position in the data frame,
if it is present in it is `select!(df, :col, :)`.
On the contrary, output column names must be unique unless `makeunique=true` is passed
(e.g. `select(df, :a, :a => :a)` or `select(df, :a, :a => sin => :a)` are not allowed).
```

Reply:
We do not support `makeunique` (at least as of now). Here is what I pushed:
```
If a collection of column names is passed to `select!` then requesting duplicate column
names in target data frame are accepted (e.g. `select!(df, [:a], :, r"a")` is allowed)
and only the first occurrence is used. In particular a syntax to move column `:col`
to the first position in the data frame is `select!(df, :col, :)`.
On the contrary, output column names of renaming, transformation and single column
selection operations must be unique, so e.g. `select!(df, :a, :a => :a)` or
`select!(df, :a, :a => sin => :a)` are not allowed.
```